### PR TITLE
Add CSR, Certificate and KeyPair

### PR DESCRIPTION
**Pull request by jderusse**
- [x] Optional argument in DistinguishedName should be optional
- [x] Add certificate formatters

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,13 @@
...
+
+[![Build Status](https://travis-ci.org/acmephp/ssl.svg?branch=master)](https://travis-ci.org/acmephp/ssl)
+[![StyleCI](https://styleci.io/repos/51226077/shield)](https://styleci.io/repos/51226077)
+
+> Note : this repository is in development
+
+This repository provide tools to generate, dump and load SSL entities
+
+- KeyPair
+- Distinguished Names
```
Comment (tgalopin) [inline README.md]: Isn't it Certificate Signing Requests ?

File: composer.json [hunk 2]
```diff
@@ -0,0 +1,52 @@
+{
+    "name": "acmephp/ssl",
+    "description": "Tools to manipulate certificates.",
+    "type": "library",
+    "license": "MIT",
+    "homepage": "https://github.com/acmephp/ssl",
+    "keywords": [
+        "certificate",
+        "letencrypt",
```
Comment (tgalopin) [inline composer.json]: This package is not linked to letsencrypt. However, it's linked to openssl, we could add it here.

File: composer.json [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    "autoload-dev": {
+        "psr-4": {
+            "AcmePhp\\Ssl\\Tests\\": "tests/"
+        }
+    },
+    "require": {
+        "php": ">=5.5.0",
+        "ext-openssl": "*",
+        "webmozart/assert": "^1.0",
+        "symfony/property-access": "^3.0"
```
Comment (tgalopin) [inline composer.json]: Do we really need this?

File: src/Certificate.php [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+    /** @var string */
+    private $subject;
+
+    /** @var string */
+    private $issuer;
+
+    /** @var bool */
+    private $selfSigned;
+    /**
+     * @var \DateTime
```
Comment (tgalopin) [inline src/Certificate.php]: Why do you use another comment style convention?

File: src/Certificate.php [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+     */
+    public function __construct(
+        $subject,
+        $issuer = null,
+        $selfSigned = true,
+        \DateTime $validFrom = null,
+        \DateTime $validTo = null,
+        $serialNumber = null,
+        array $subjectAlternativeNames = []
+    ) {
```
Comment (tgalopin) [inline src/Certificate.php]: I think we should validate data with Assert here

File: src/DistinguishedName.php [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+     * @param string $countryName
+     * @param string $stateOrProvinceName
+     * @param string $localityName
+     * @param string $organizationName
+     * @param string $organizationalUnitName
+     * @param string $emailAddress
+     * @param array  $subjectAlternativeNames
+     */
+    public function __construct(
+        $commonName,
```
Comment (tgalopin) [inline src/DistinguishedName.php:58]: I'm not sure every parameter is required in OpenSSL CSR, I think we should find out which one is really.

File: src/Exception/CSRSigninException.php [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace AcmePhp\Ssl\Exception;
+
+/**
+ * @author Jérémy Derussé <jeremy@derusse.com>
+ */
+class CSRSigninException extends AcmeSslException
```
Comment (tgalopin) [inline src/Exception/CSRSigninException.php]: Can you add in comment when exceptions are thrown? It's useful for debugging.

File: src/Factory/KeyPairFactory.php [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+use AcmePhp\Ssl\KeyPair;
+use AcmePhp\Ssl\PrivateKey;
+use AcmePhp\Ssl\PublicKey;
+
+/**
+ * Crate KeyPair entities from PEMs.
+ *
+ * @author Jérémy Derussé <jeremy@derusse.com>
+ */
+class KeyPairFactory
```
Comment (tgalopin) [inline src/Factory/KeyPairFactory.php]: I don't like this kind of factories. Either we use directly the KeyPair or we create an interface that the factory implements. The aim of the Factory pattern is to be used with Strategy so anyone can replace it implementing the interface themselves. IMO without an interface it's pointless.

File: composer.json (hunk 3 above)
Comment (jderusse) [inline composer.json]: You are right, it's overkill

File: src/Certificate.php (hunk 4 above)
Comment (jderusse) [inline src/Certificate.php]: My ide :/

File: src/DistinguishedName.php (hunk 6 above)
Comment (jderusse) [inline src/DistinguishedName.php:58]: Yes !

was a copy/paste from the core. But it's time to change it

File: src/Factory/KeyPairFactory.php (hunk 8 above)
Comment (jderusse) [inline src/Factory/KeyPairFactory.php]: I don't like having a `new` outside a factory. But here, the factory is useless... :/

Comment (tgalopin) [inline src/Factory/KeyPairFactory.php]: Why don't you like new? Generally speaking, the simpler is usually the better, creating a factory like this one seems overkill IMO.

File: .gitignore [hunk 9]
```diff
@@ -0,0 +1,4 @@
+composer.lock
+vendor/
+.php_cs.cache
+build/
```
Comment (jderusse) [inline .gitignore]: Missing end line

File: src/Factory/KeyPairFactory.php (hunk 8 above)
Comment (jderusse) [inline src/Factory/KeyPairFactory.php]: To respecr SRP. it's the job of factories to create object.

Thee factory should be used in the repositories too. 
But I also thing it's over complicated in this case.

Comment (tgalopin) [inline src/Factory/KeyPairFactory.php]: Single Responsibility Principle works when you have more than a couple of line of code IMO.

File: README.md [hunk 10]
```diff
@@ -1,2 +1,13 @@
...
+- Certificates
```
Comment (tgalopin) [inline README.md]: Missing end line :)

File: src/Certificate.php (hunk 5 above)
Comment (jderusse) [inline src/Certificate.php]: done

File: .gitignore (hunk 9 above)
Comment (jderusse) [inline .gitignore]: done

File: README.md (hunk 10 above)
Comment (jderusse) [inline README.md]: done

File: src/DistinguishedName.php (hunk 6 above)
Comment (tgalopin) [inline src/DistinguishedName.php:58]: According to https://en.wikipedia.org/wiki/Certificate_signing_request#Structure, `$subjectAlternativeNames` is not required (I think we should set it as empty array by default) but all other fields are.

**Comment by jderusse** [comment]
I add formatters and new entities (CertificateRequest, CertificateResponse, CertificateChain) to works with concrete objects instead of PEM string or resources.

The common name is the only one requirement, to generate a valid CSR (as the official python client do)

File: src/Formatter/CombinedFormatter.php [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+        $this->keyPairFormatter = $keyPairFormatter;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function format(CertificateResponse $certificateResponse)
+    {
+        return implode(
+            '',
```
Comment (tgalopin) [inline src/Formatter/CombinedFormatter.php]: Shouldn't this be a new line? I think we shouldn't rely on the OpenSSL ending new line at files bottom

File: src/Formatter/FullChainFormatter.php [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+        $this->certificateFormatter = $certificateFormatter;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function format(CertificateResponse $certificateResponse)
+    {
+        return implode(
+            '',
```
Comment (tgalopin) [inline src/Formatter/FullChainFormatter.php]: Shouldn't be this a new line?

File: src/Formatter/IssuerChainFormatter.php [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function format(CertificateResponse $certificateResponse)
+    {
+        $certificateChain = $certificateResponse->getIssuerCertificateChain();
+        $PEMs = [];
+        do {
+            $PEMs[] = $certificateChain->getCertificate()->getPEM();
```
Comment (tgalopin) [inline src/Formatter/IssuerChainFormatter.php]: Are we sure we always will have a first certificate? I think in the case of self-signed and self-trusted certificates, we won't.

File: src/ParsedCertificate.php [hunk 14]
```diff
@@ -0,0 +1,135 @@
...
+namespace AcmePhp\Ssl;
+
+use Webmozart\Assert\Assert;
+
+/**
+ * Represent the content of a parsed certificate.
+ *
+ * @author Jérémy Derussé <jeremy@derusse.com>
+ */
+class ParsedCertificate
```
Comment (tgalopin) [inline src/ParsedCertificate.php:21]: What do you think of  extending Certificate base classe? Seems logical at the first glimpse but I'm not sure

**Comment by tgalopin** [comment]
That's really great, it will be a very stable basis for the other packages :) !

File: src/Formatter/CombinedFormatter.php (hunk 11 above)
Comment (jderusse) [inline src/Formatter/CombinedFormatter.php]: You're right.

File: src/Formatter/IssuerChainFormatter.php (hunk 13 above)
Comment (jderusse) [inline src/Formatter/IssuerChainFormatter.php]: I didn't like my first design of the CertificateChain entity.
I'll refactor it and fix this case too

File: src/ParsedCertificate.php (hunk 14 above)
Comment (jderusse) [inline src/ParsedCertificate.php:21]: Prefere the composition in this case

**Pull request merged**